tools/x509/openssl.py

```diff
@@ -0,0 +1,124 @@
+"""
+Wrapper around openssl providing several methods to manage X.509 certificates.
+"""
+
+import contextlib
+import subprocess
+import tempfile
+from os import path
+
+
+def _config(common_name, subject_alternative_names):
+    """
+    Returns config for creating x.509 v3 certificate using openssl.
+    :param common_name: Common name of the subject owning the certificate
+    :param subject_alternative_names: Subject alternative names of the subject owning the certificate
+    :return: openssl config as a string
+    """
+    alt_names = "\n".join((f"DNS.{i + 1} = {san}" for (i, san) in enumerate(subject_alternative_names)))
+    return f"""
+[req]
+distinguished_name = req_distinguished_name
+req_extensions = v3_req
+prompt = no
+
+[req_distinguished_name]
+CN = {common_name}
+
+[v3_req]
+keyUsage = critical,keyEncipherment, dataEncipherment, digitalSignature
+extendedKeyUsage = critical,serverAuth,clientAuth,emailProtection
+basicConstraints = critical,CA:FALSE
```

Comment:
This generates certificates with a wide range of usages. This is fine for tests, but we ought to never use them in production.

In my understanding, `CA` needs to be `true` for the ca certificate. It seems to work regardless — do you know why?

Reply:
I have no idea, the config is taken from `koji-osbuild`. I'm open to all suggestions.

Reply:
It needs to be like this when using the same config for all certificates. On a production system, you'd *only* allow `serverAuth` for the server cert, only `clientAuth` for a client cert, etc. (Again, it is fine for testing as is.)

I'm not sure why `CA:FALSE` works. The man page `X509V3_CONFIG(5)` says:

> A CA certificate must include the basicConstraints value with the CA field set to TRUE.